pkg/handlers/generic/lifecycle/addons/helmaddon.go

```diff
@@ -88,29 +133,42 @@ func (a *helmAddonApplier) Apply(
 		)
 	}
 
+	if applyOpts.valueTemplater != nil {
+		values, err = applyOpts.valueTemplater(cluster, values)
+		if err != nil {
+			return fmt.Errorf("failed to template Helm values: %w", err)
+		}
+	}
+
+	targetCluster := cluster
+	if applyOpts.targetCluster != nil {
+		targetCluster = applyOpts.targetCluster
+	}
+
 	chartProxy := &caaphv1.HelmChartProxy{
 		TypeMeta: metav1.TypeMeta{
 			APIVersion: caaphv1.GroupVersion.String(),
 			Kind:       "HelmChartProxy",
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			Namespace: cluster.Namespace,
-			Name:      a.config.defaultHelmReleaseName + "-" + cluster.Name,
+			Namespace: targetCluster.Namespace,
+			Name:      fmt.Sprintf("%s-%s", a.config.defaultHelmReleaseName, clusterUUID),
 		},
 		Spec: caaphv1.HelmChartProxySpec{
 			RepoURL:   a.helmChart.Repository,
 			ChartName: a.helmChart.Name,
 			ClusterSelector: metav1.LabelSelector{
-				MatchLabels: map[string]string{clusterv1.ClusterNameLabel: cluster.Name},
+				MatchLabels: map[string]string{clusterv1.ClusterNameLabel: targetCluster.Name},
 			},
 			ReleaseNamespace: a.config.defaultHelmReleaseNamespace,
 			ReleaseName:      a.config.defaultHelmReleaseName,
```

Comment:
The CA Deployment selectors end up being generic
```
    selector:
      matchLabels:
        app.kubernetes.io/instance: cluster-autoscaler
        app.kubernetes.io/name: clusterapi-cluster-autoscaler
```

~Do we want the `ReleaseName` to also be `fmt.Sprintf("%s-%s", a.config.defaultHelmReleaseName, clusterUUID),`?~ 

Error in CAAPH log:
```
E0812 19:52:49.197830       1 controller.go:329] "Reconciler error" err="releaseContent: Release name is invalid: cluster-autoscaler-01914825-cd57-7cc1-b28c-528d1348b3eb" controller="helmreleaseproxy" controllerGroup="addons.cluster.x-k8s.io" controllerKind="HelmReleaseProxy" HelmReleaseProxy="default/cluster-autoscaler-dkoshkin-super-long-cluster-name-that-swgncj" namespace="default" name="cluster-autoscaler-dkoshkin-super-long-cluster-name-that-swgncj" reconcileID="b5634b36-3a43-4c68-8198-715cf9a5710b"
```

Reply:
Aaagggh must have reverted too much from previous changes 😞

I'll fix the release name tomorrow. Can you just list what you think I need to double check and I'll add tests for it too. I can't think straight anymore 😅

Reply:
Try with latest commit...

Reply:
> I also don't understand where the HRP names are coming from, the ClusterName is `dkoshkin-super-long-cluster-name-that-should-work`, is the CAAPH controller trimming that and including that in the HRP name?

Yes, exactly. Trimming to be under 63 chars I guess and using `GenerateName` to get uniqueness.

Reply:
Thanks for the fix, looking great! The only test case I can think of thats missing is checking the `spec.selector` field on the CA Delployment.
```
 $ k get deploy -o yaml
 apiVersion: v1
items:
- apiVersion: apps/v1
  kind: Deployment
  metadata:
    annotations:
      deployment.kubernetes.io/revision: "1"
      meta.helm.sh/release-name: ca-019148ad-b788-71f2-81a4-ae3210c3520d
      meta.helm.sh/release-namespace: default
    creationTimestamp: "2024-08-12T22:21:51Z"
    generation: 1
    labels:
      app.kubernetes.io/instance: ca-019148ad-b788-71f2-81a4-ae3210c3520d
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: clusterapi-cluster-autoscaler
      helm.sh/chart: cluster-autoscaler-9.37.0
    name: cluster-autoscaler-019148ad-b788-71f2-81a4-ae3210c3520d
    namespace: default
...
spec:
    selector:
      matchLabels:
        app.kubernetes.io/instance: ca-019148ad-b788-71f2-81a4-ae3210c3520d
        app.kubernetes.io/name: clusterapi-cluster-autoscaler
```

And for anyone else from the future wondering why the `addonName` variable was changed from `cluster-autoscaler-` to `ca-`, its because of https://github.com/helm/helm/issues/6006.
With the generated UUID a Helm release a name like `cluster-autoscaler-0191489a-82b3-758e-a03d-ea7e5e803c90` is 55 characters, longer than Helm's 53 char limit, which explains the `Release name is invalid` error I was seeing.

Reply:
Added test case to ensure the label has a UUID at the end.